Added PSR-6 adapter

kynx opened the pull request:
This PR address #46, adding a [PSR-6](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-6-cache.md) wrapper.

Some key points:

* Not all storage adapters are supported, due the PSR-6 requirement that individual cache items must support separate TTLs, set on save. This rules out Dba, Filesystem, Memory, Session and Redis < v2
* PSR-6 [Error Handling](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-6-cache.md#error-handling) requires that most exceptions thrown by the storage adapter be suppressed. I've provided an `ExceptionLogger` so these can be saved to a PSR-3 compatible logger. Some documentation is probably in order!
* There are some useful integration tests at https://github.com/php-cache/integration-tests. I haven't got around to incorporating those yet - only discovered them yesterday.

Happy New Year!

Nyholm commented:
The implementation looks good. There were some small bugs thou.. Say you do like this: 
```php
$item = $pool->getItem('key');
$item->set('foo');
$pool->saveDeferred($item);

$item = $pool->getItem('key');
$item->set('bar');
// No save 

$item = $pool->getItem('key');
$item->get(); // should be 'foo'

```



I added some integration tests to this PR: https://github.com/kynx/zend-cache/pull/1

Nyholm commented on src/Psr/CacheItem.php:
```diff
@@ -0,0 +1,130 @@
+     */
+    public function get()
+    {
+        return $this->value;
```
You need to check if isHit before you return anything

kynx replied:
The value is set to `null` in the constructor if the item isn't a hit.

Nyholm commented on src/Psr/CacheItem.php:
```diff
@@ -0,0 +1,130 @@
+     */
+    public function isHit()
+    {
+        return $this->isHit;
```
Say that you pull something from cache, waits 10 seconds and then the cache expires. Then isHit() should be false. 

This function needs to verity the expiration time

kynx replied:
Are you 100% sure about this? The docblock states "Confirms if the cache item lookup resulted in a cache hit", and I can't find anything in the rest of the spec that specifically talks about expiration post-lookup.

Doing it any other way with Zend Cache would not be reliable: TTL is not accessible on fetch per-item - we only have the default TTL for the storage adapter.

Nyholm replied:
Sorry, you are correct. I miss-remembered this paragraph: 

> Hit - A cache hit occurs when a Calling Library requests an Item by key and a matching value is found for that key, and that value has not expired, and the value is not invalid for some other reason. Calling Libraries SHOULD make sure to verify isHit() on all get() calls.

It says "Calling Libraries SHOULD make sure to verify isHit() on all get() calls.". I remembered it as  "Implementing Libraries... ". 

So a there is not need to change `$this->isHit = false` at any time in your implementation.

Nyholm commented on src/Psr/CacheItemPoolAdapter.php:
```diff
@@ -0,0 +1,287 @@
+            throw new InvalidArgumentException('$item must be an instance of ' . CacheItem::class);
+        }
+
+        $this->validateKey($item->getKey());
```
No *need* to validate the key. Calling libraries should never instantiate a `CacheItem`. So the key is already checked by `getItems`.

Nyholm commented on src/Psr/CacheItemPoolAdapter.php:
```diff
@@ -0,0 +1,287 @@
+        }
+
+        $key = $item->getKey();
+        $this->validateKey($key);
```
No *need* to validate the key. Se comment above.

marc-mabe commented on src/Psr/CacheItem.php:
```diff
@@ -0,0 +1,130 @@
+        }
+
+        $this->expiration = new \DateTime();
+        $this->expiration->add($interval);
```
I'm not sure if this is safe for all time zones as in dst time zones there are times ambiguous.
see https://3v4l.org/BnPd0 (check the diff)
-> Can we force the time zone as UTC ?

marc-mabe commented on src/Psr/CacheItem.php:
```diff
@@ -0,0 +1,130 @@
+            throw new InvalidArgumentException(sprintf('Invalid $time "%s"', gettype($time)));
+        }
+
+        $this->expiration = new \DateTime();
```
`use DateTime;`

marc-mabe commented on src/Psr/CacheItem.php:
```diff
@@ -0,0 +1,130 @@
+    {
+        if ($time instanceof DateInterval) {
+            $interval = $time;
+        } elseif ((int) $time == $time) {
```
Shouldn't that be `is_int($time)`?

marc-mabe commented on src/Psr/CacheItem.php:
```diff
@@ -0,0 +1,130 @@
+            $interval = $time;
+        } elseif ((int) $time == $time) {
+            $interval = new DateInterval('PT' . $time . 'S');
+        } else {
```
I'm missing the handling of `null` to get the default TTL defined by adapter

marc-mabe commented on src/Psr/CacheItemPoolAdapter.php:
```diff
@@ -0,0 +1,287 @@
+
+        // we've got to be able to set per-item TTL on write
+        $capabilities = $storage->getCapabilities();
+        if (!($capabilities->getStaticTtl() && $capabilities->getMinTtl())) {
```
I think the check for `&& $capabilities->getMinTtl()` should be removed as this only describes that it's possible to store an item without expiration but not that the adapter doesn't support expiration.

kynx replied:
Heh. I actually nicked that from your code about 2/3 of the way down https://github.com/zendframework/zf2/pull/5386 :smile: 

Redis actually sets `staticTtl = true` and `minTtl = 0` if version < 2. Not entirely sure what that means, but decided to play safe.

marc-mabe replied:
Oh wow you are totally right - I shouldn't comment with my head somewhere else.
-> Please ignore my previous comment

marc-mabe commented on src/Psr/CacheItemPoolAdapter.php:
```diff
@@ -0,0 +1,287 @@
+            if ($this->storage instanceof ClearByNamespaceInterface && $namespace) {
+                $cleared = $this->storage->clearByNamespace($namespace);
+            } else {
+                $cleared = $this->storage->flush();
```
Is there any description about that difference in PRS-6?

kynx replied:
No, the spec has no concept of namespaces. I'm doing this to avoid clobbering cache items outside the namespace set up for the storage adapter, if it supports them.

marc-mabe replied:
mh PSR-6 has no namespace definition but the definition of a pool could be interpreted for us as a collection of items of one namespace of the same storage.
From this POV it would be OK, even to flush the hole cache storage as fallback would effect items of another pool but for a cache and the spec there is no guarantee that items exists as long as they are not deleted or expired. So it only effects performance of the different pool.

:+1:

marc-mabe commented on src/Psr/CacheItemPoolAdapter.php:
```diff
@@ -0,0 +1,287 @@
+        $items = $this->getItems([$key]);
+
+        return array_pop($items);
+    }
```
As single items are used much more in user land than a bunch of items I prefer to remove the small overhead of wrapping a single item call into a bunch of one item.

marc-mabe commented on src/Psr/CacheItemPoolAdapter.php:
```diff
@@ -0,0 +1,287 @@
+
+    private function validateKey($key)
+    {
+        if (!is_string($key) || preg_match('#[{}()/\\\\@:]#', $key)) {
```
We should find a way to validate the key by adapter as some adapters have additional limitations.

kynx replied:
I'm catching and re-throwing any `InvalidArgumentException`s thrown by the storage adapter to cover this. It's a bit blunt - they could be thrown outside `AbstractAdapter::normalizeKey()`, though I couldn't spot this in the code paths I'm using. But if that's a worry it wouldn't be hard to cut-and-paste the `normalizeKey()` logic.

marc-mabe commented:
Hi @kynx thanks for your work!

I have added some notes from a quick look.
What I don't understand is why did you addad the PSR-3 `ExceptionLogger`? In my opinion this has nothing to do with the PSR-6 implementation and we already have a storage adapter plugin you know where you can handle all thrown exceptions as you like.

kynx commented:
@marc-mabe Many thanks for the feedback.

The `ExceptionLogger` is there for a couple of reasons:

1. The [spec](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-6-cache.md#error-handling) says: "An Implementing Library SHOULD log such errors or otherwise report them to an administrator as appropriate."
2. There's a chance a user will attach an `ExceptionHandler` plugin that has `'throw_exceptions' => false`. That would screw up the boolean return values of most methods that talk to storage if it failed. So the `ExceptionLogger` will barf if it finds another `ExceptionHandler` plugin attached.

I didn't want to completely rule out users handling exceptions however they wanted, but I did imagine that any Service class would add the logger by default - even if it defaulted to a `NullLogger`. But I'll ditch it if you want.

Let me know if there's more to do with `validateKey()` and the `clear()` stuff.

kynx commented:
@Nyholm Many thanks for the feedback and integration tests. I've mocked out the adapter in those now and made sure they all pass :smiley:

Nyholm commented:
> @Nyholm Many thanks for the feedback and integration tests. I've mocked out the adapter in those now and made sure they all pass :smiley:

You are welcome. 

I see that your mock preforms better than an actual adapter. When I test this with the redis adapter I fail a few tests. The redis adapter does not preserve datatypes. Consider this test:

```php
$item = $this->cache->getItem('key');
$item->set(5); // Integer 5
$this->cache->save($item);

$item = $this->cache->getItem('key');
$this->assertTrue(5 === $item->get(), 'Wrong data type. If we store an int we must get an int back.');
```

Also, this just crossed my mind. This is not a part of PSR-6 as far as I know. But what if you do like this: 
```php
$item = $this->cache->getItem('key');
$item->set('foo');
$item->expiresAfter(3600);
$this->cache->save($item);

// 10 minutes later in an other request
$item = $this->cache->getItem('key');
$item->set('bar');
$this->cache->save($item);
```
I'm not too sure how the storage options work. For how long will the item be saved? 3000 more seconds? 3600 seconds? Or infinity?

kynx commented:
Yeah, I noticed that fail with Redis too. Do you happen to know if this is a general issue with how Redis serializes data, or if it's specific to Zend Cache? 

OT, but it would be useful to provide a mechanism in the integration tests for marking certain tests skipped - maybe be just an associative array of methodName => reason that gets checked at the start of each test.

Regarding re-saving an item after TTL has been set: the default TTL configured for the storage would be used in this case. I don't think any implementation could do otherwise reliably - another process may have cleared the cache, it might be unavailable, etc.

marc-mabe commented:
> The ExceptionLogger is there for a couple of reasons:
>
>    The spec says: "An Implementing Library SHOULD log such errors or otherwise report them to an administrator as appropriate."
>    There's a chance a user will attach an ExceptionHandler plugin that has 'throw_exceptions' => false. That would screw up the boolean return values of most methods that talk to storage if it failed. So the ExceptionLogger will barf if it finds another ExceptionHandler plugin attached.

The spec makes a recommendation to log errors or report them but nobody as the calling library can actually handle errors. Also I don't see a difference where your `ExceptionLogger` makes it in any case simpler to log errors. `zend-cache` already makes it very simple to handle errors.

> Yeah, I noticed that fail with Redis too. Do you happen to know if this is a general issue with how Redis serializes data, or if it's specific to Zend Cache? 

Redis has no serialize feature, it simply gets requires a string and PHP converts any given value into a string at this point.

From spec
```
Implementing libraries MUST support all serializable PHP data types, including:

    Strings - Character strings of arbitrary size in any PHP-compatible encoding.
    Integers - All integers of any size supported by PHP, up to 64-bit signed.
    Floats - All signed floating point values.
    Boolean - True and False.
    Null - The actual null value.
    Arrays - Indexed, associative and multidimensional arrays of arbitrary depth.
    Object - Any object that supports lossless serialization and deserialization such that $o == unserialize(serialize($o)). Objects MAY leverage PHP's Serializable interface, __sleep() or __wakeup() magic methods, or similar language functionality if appropriate.
```

That means you have to make sure the adapter (or the serializer plugin) serializes data (reported by `getCapabilities()->getSupportedDataTypes()`) and if not you need to `un/serialize` the data by hand to match the spec.

kynx commented:
I've dropped `ExceptionLogger`.

My holidays are over. I no longer have the time to look into the hairy details of Redis serialisation. PR is welcome.

Thanks.

Nyholm commented:
> If the Redis adapter doesn't handle serialising/unserialising data types correctly, isn't that a problem for the Redis adapter? I don't really see why my PRS-6 thingy should be compensating for the underlying adapter getting it wrong.

Yes, that is possible. It depends how you want to architecture your Pool, Clients and adapters. If you decide to push that problem to the adapters I would recommend doing integration tests with all adapters instead of a mock, just to make sure the adapters work as well.

aequasi commented on src/Psr/CacheItem.php:
```diff
@@ -0,0 +1,139 @@
+    public function __construct($key, $value, $isHit)
+    {
+        $this->key = $key;
+        $this->value = $isHit ? $value : null;
```
null is a valid value for PSR-6

kynx replied:
Yes, which is why you need to check isHit() - afaik the value should be null if its not a hit.

marc-mabe commented:
@kynx the redis adapter doesn't handle it wrong. It does what it suppose to do - sending data to redis in a format that redis supports and that's a string only. If you want more you have to add the serializer plugin. That's because not every one need this support and do serializing even if not required would be an useless overhead. That's how and why zend-cache is designed. So if the underlying adapter doesn't support all requirements from psr-6 this wrapper need to implement it on top.

kynx commented:
@marc-mebe yes I had another l

kynx commented:
Darn phone! Yes, I had another look at this and am completely wrong :(

I'm currently putting together integration tests for each adapter. Once that's done I can start working around the differences.

Nyholm commented:
About skipping tests: https://github.com/php-cache/integration-tests/pull/16

kynx commented:
I've added integration tests for each of the storage adapters and am now using the `Serializer` plugin where there are unsupported data types. Can't believe I missed that in the docs to start with :|

I struggled getting phpunit working on a windows VM for too many hours before giving up, so haven't been able to run `test/Psr/WinCacheIntegrationTest.php`. If anyone has a working setup I'd appreciate a check.

Similarly I haven't got the ZendServer adapter tests running yet. I've added a `'minTtl' => 1` to `AbstractZendServer` adapter options there. Hope that's OK - it's in the documentation and just looks like an oversight. 

I've also added some documentation to docs/book.

Looks like `StorageCacheAbstractServiceFactoryTest` is gonna fail due to a change in the ServiceManager dep. Cross my fingers everything else is OK.

marc-mabe commented on src/Psr/CacheItemPoolAdapter.php:
```diff
@@ -0,0 +1,339 @@
+                    get_class($storage)
+                ));
+            }
+        }
```
@kynx I don't thing it's a good idea to actually change the underlying adapter from within a wrapper class. I would be better to hold this information in an object property on initializing the object and directly calling un-/serialize in the wrapper.

kynx replied:
Oh, OK. I can see your point.

marc-mabe commented on src/Storage/Adapter/AbstractZendServer.php:
```diff
@@ -216,6 +216,7 @@ protected function internalGetCapabilities()
                         'resource' => false,
                     ],
                     'supportedMetadata'  => [],
+                    'minTtl'             => 1,
```
please make a separate PR against `master` as this is a bug fix.

kynx replied:
OK, will do.

marc-mabe commented:
@kynx @Nyholm What happens if an deferred item expires before pushing it to storage? Is it required to check TTL there, too?

kynx commented:
It's a good question, and probably one for the FIG ML. And come to that, what if it expires before you call save().

My feeling is that expiry dates from when the item is stored, not when the value is set. But then I'm just a lazy programmer :smile:  

@Nyholm has another interesting one over at https://github.com/php-cache/integration-tests/issues/25. I might be wrong, but implementing anything other than his option C would be non-trivial over here.

But I think it's important to remember these are edge cases. In the 4 years it to the PSR to pass no one thought of them, so I'm not *too* worried.

kynx commented:
Regarding expiry and deferred items, I guess the relevant part of the specs are:

> TTL - The Time To Live (TTL) of an item is the amount of time between when that item is stored and it is considered stale. The TTL is normally defined by an integer representing time in seconds, or a DateInterval object.

This makes it clear that the TTL is counted from when the item is stored. But...

> Expiration - The actual time when an item is set to go stale. This it typically calculated by adding the TTL to the time when an object is stored, but may also be explicitly set with DateTime object.

I'm not convinced the intention was to treat `expiresAt()` and `expiresAfter()` differently, but will ask the author(s) to explain.

kynx commented:
I've had a re-think about `expiresAt()` and refactored it so expired items aren't pushed to storage, and aren't a hit. I've also changed the serialization so it's now done by the pool, not be the Serializer plugin.

marc-mabe commented on src/Psr/CacheItemPoolAdapter.php:
```diff
@@ -0,0 +1,360 @@
+            if (is_null($itemTtl) || $itemTtl > 0) {
+                $saved = $this->storage->setItem($item->getKey(), $value);
+                // saved items are a hit? see integration test CachePoolTest::testIsHit()
+                $item->setIsHit(true);
```
You should check `$saved` before blindly describing the item has hit.

kynx replied:
Good point.

marc-mabe commented on src/Psr/CacheItemPoolAdapter.php:
```diff
@@ -0,0 +1,360 @@
+            }
+
+            // don't save items that have already expired
+            if (is_null($itemTtl) || $itemTtl > 0) {
```
This should result in exactly the same behavior : `if ($itemTtl >= 0) {`

kynx replied:
See next one - if you've used `expiresAt()` and that date has since passed, `getTtl() <= 0`.

marc-mabe commented on src/Psr/CacheItemPoolAdapter.php:
```diff
@@ -0,0 +1,360 @@
+            $itemTtl = $item->getTtl();
+            if ($itemTtl > 0) {
+                $options->setTtl($itemTtl);
+            }
```
Please correct me if I'm wrong but doesn't `0` mean the item never expires and that `null` mean no TTL defined = use the default TTL of the adapter?

kynx replied:
In PSR-6-land null means use default TTL. I've added the `getTtl()` method for testing whether an item has expired and to use in this case to set the value when storing. For example,

```
$item->expiresAt(new DateTime(null));
echo $item->getTtl();
// outputs 0 - item has expired
```

marc-mabe commented on src/Psr/CacheItemPoolAdapter.php:
```diff
@@ -0,0 +1,360 @@
+
+        // deferred items should always be a 'hit'
+        $item->setIsHit(true);
+        $this->deferred[$item->getKey()] = $item;
```
Even if the item has been expired already ?

kynx replied:
Yeah: `isHit()` checks expiry for items that have had `expiresAt()` set. We don't know that for sure until the time when it's checked.

marc-mabe replied:
ah ok I got it

marc-mabe commented on composer.json:
```diff
@@ -16,10 +16,13 @@
-        "zendframework/zend-eventmanager": "dev-develop as 2.7.0"
+        "zendframework/zend-eventmanager": "dev-develop as 2.7.0",
+        "psr/cache": "^1.0",
+        "psr/log": "^1.0"
```
the `psr/log` dependency should be removed

kynx replied:
Will do - need to give travis a kick anyway ;)

marc-mabe commented on src/Psr/CacheItem.php:
```diff
@@ -0,0 +1,145 @@
+    private $value;
+    private $isHit = false;
+    private $expiration = null;
+    private $ttl = null;
```
Please can you add code comments for each of the properties - thanks

kynx replied:
Will do.

marc-mabe commented on src/Psr/CacheItem.php:
```diff
@@ -0,0 +1,145 @@
+        $ttl = $this->ttl;
+        if ($this->expiration instanceof DateTime) {
+            $now = new DateTime('now', $this->tz);
+            $ttl = $this->expiration->getTimestamp() - $now->getTimestamp();
```
If you sore the expire time as integer you can calculate the TTL here much faster with `$ttl = $this->expiration - time()`

kynx replied:
We need to recalculate so we can find out whether a deferred item has been expired in the meantime.

marc-mabe commented on src/Psr/CacheItem.php:
```diff
@@ -0,0 +1,145 @@
+    public function isHit()
+    {
+        $ttl = $this->getTtl();
+        return $this->isHit && (is_null($ttl) || $ttl > 0);
```
Minor performance note: `return $this->isHit && ($ttl = $this->getTtl() === null || $ttl > 0)`

marc-mabe commented on test/Psr/ApcIntegrationTest.php:
```diff
@@ -0,0 +1,72 @@
+
+        // needed on test expirations
+        $this->iniUseRequestTime = ini_get('apc.use_request_time');
+        ini_set('apc.use_request_time', 0);
```
Does `apc.use_request_time = 1` break PSR-6 specification? So so an exception should be thrown on instantiating in this case :(

kynx replied:
Oh lordy, yes it does. Should've paid more attention to that when I was cut-and-pasting the tests. I've added a check to the constructor and a test.

kynx commented:
Looks like CI is going to fail until #64 is in.

samsonasik commented:
setUp() function in tests should use protected modifier in favor of https://github.com/php-cache/integration-tests/pull/31

marc-mabe commented on src/Psr/CacheItem.php:
```diff
@@ -0,0 +1,170 @@
+    public function isHit()
+    {
+        $ttl = $this->getTtl();
+        return $this->isHit && ($ttl === null || $ttl > 0);
```
You should avoid calling `getTtl` if the item is already marked as `isHit === false`

marc-mabe commented on src/Psr/CacheItem.php:
```diff
@@ -0,0 +1,170 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function expiresAt($expiration = null)
```
the spec has no default value here `public function expiresAt($expiration);`

marc-mabe commented on src/Psr/CacheItemPoolAdapter.php:
```diff
@@ -0,0 +1,394 @@
+                    if ($this->serializeValues) {
+                        $value = unserialize($value);
+                    }
+                };
```
`getItem` should be used with the second argument by reference instead of calling `hasItem` and `getItem` to avoid concurrency issues and for performance reasons as some adapters doesn't natively support `hasItem` (which will be emulated with `getItem`).

Also errors in `unserialize` needs to be handled (see https://github.com/zendframework/zend-serializer/blob/master/src/Adapter/PhpSerialize.php#L71)

```php
// ...
$value = $this->storage->getItem($key, $isHit);
if ($isHit && $this->serializeValues) {
    if ($value === static::$serializedFalse) {
        $value = false;
    } elseif ($value = unserialize($value) === false) {
       // TODO: what should happen if unserialization failed?
    }
}
// ...
```

kynx replied:
Grrr... Yeah, that makes sense, but prophesy sure don't like args-by-reference with `getItem()`. Still looking at what I can do for the tests.

marc-mabe commented on src/Psr/CacheItemPoolAdapter.php:
```diff
@@ -0,0 +1,394 @@
+
+            foreach ($cacheItems as $key => $value) {
+                if ($this->serializeValues) {
+                    $value = unserialize($value);
```
Same here : errors in `unserialize` needs to be handled

kynx commented:
A little bit of `Reflection` was required to test unserialization, which feels like a hack, but there we go - prophesy doesn't let me do arg-by-ref.

If a value can't be `unserialize`d it will simply show up as a cache miss. I can't see a straightforward way of reporting the problem to the user that doesn't break PSR-6. But - unless something else is writing to the same storage adapter - I can't see it being more than a sporadic error. Let me know if you've got other ideas.

@samsonasik - thanks for the heads up. Saw your https://github.com/php-cache/integration-tests/pull/31 - fixed here now too.

kynx commented:
If anyone's interested in Symfony's take on this, check out https://github.com/symfony/symfony/pull/17408

samsonasik commented on composer.json:
```diff
@@ -16,10 +16,12 @@
         "zendframework/zend-stdlib": "~2.5",
         "zendframework/zend-servicemanager": "dev-develop as 2.7.0",
-        "zendframework/zend-eventmanager": "dev-develop as 2.7.0"
+        "zendframework/zend-eventmanager": "dev-develop as 2.7.0",
```
zendframework/zend-eventmanager ^2.6 || ^3.0 ? /cc @weierophinney

marc-mabe replied:
@samsonasik That's not part of this PR - see #64

aequasi commented on composer.json:
```diff
@@ -16,10 +16,12 @@
     },
     "require-dev": {
         "zendframework/zend-serializer": "dev-develop as 2.6.0",
+        "cache/integration-tests": "dev-master",
```
This may want to change to @stable or `^0.7`

kynx commented:
I've rebased against develop and the tests are passing :)

marc-mabe commented on doc/book/psr6.md:
```diff
@@ -0,0 +1,132 @@
+The PSR-6 specification requires that the underlying storage support time-to-live (TTL), which is set when the
+item is saved. For this reason the following adapters cannot be used: `Dba`, `Filesystem`, `Memory` and `Session`. The 
+`XCache` adapter calculates TTLs based on the request time, not the time the item is actually persisted, which means 
+that it also cannot be used.
```
I think this is confusing because:
- The filesystem and memory adapter supports TTL but it's based on a stored last-modification-time + runtime TTL setting
  - I'll open an issue for the Memory adapter to change this because this should be a simple change and could also improve performance but it's a BC break.
  - For the filesystem adapter this would be worst because it would be required to read the file before to know it's expired.
- The XCache adapter is not the only adapter calculating TTLs based on request time. APC/APCu also have this behavior if `apc.use_request_time` which is enabled be default.

marc-mabe commented on doc/book/psr6.md:
````diff
@@ -0,0 +1,132 @@
+```
+
+Note that `throw_exceptions` should always be `true` (the default) or you will not get the correct return values from
+calls on the pool such as `save()`.
````
Why?

kynx replied:
Because we catch the exceptions and return false from these methods in case of errors, as per the spec. If they're not thrown we'll always return true. I'll clarify the docs.

marc-mabe commented on doc/book/psr6.md:
```diff
@@ -0,0 +1,132 @@
+
+Not all adapters can natively store all these types. For instance, Redis stores booleans and integers as a string. Where 
+this is the case *all* values will be automatically run through `serialize()` on save and `unserialize()` on get: you 
+do not need to use a `Zend\Cache\Storage\Plugin\Serializer` plugin.
```
... but you can if you like to change the serializer to igbinary.

marc-mabe commented on src/Psr/CacheItemPoolAdapter.php:
```diff
@@ -0,0 +1,419 @@
+                $notSaved[] = $item;
+            }
+        }
+        $this->deferred = $notSaved;
```
TODO for later: This could be improved a lot by using the `setItems` method instead but it would increase complexity.

kynx replied:
Yeah, I've thought about this a bit. For optimal performance I'd have to group the items by TTL and send each as a batch after setting the options on the adapter. Maybe one for later.

On a related note we should probably have some kind of (configurable) limit on the number of items that can be deferred before commit() is automatically called to avoid OOM situations in cache-warmers.

marc-mabe replied:
Yes this can be done later!

marc-mabe commented on src/Psr/CacheItemPoolAdapter.php:
```diff
@@ -0,0 +1,419 @@
+        }
+
+        return ['value' => $value, 'isHit' => true];
+    }
```
This seems ugly to me. Can we change that to:
```php
private function unserialize($value, & $success)
{
    if ($value === static::$serializedFalse) {
        $success = true;
        return false;
    }

    $value = unserialize($value);
    if ($value === false) {
        $success = false;
        return null;
    }

    $success = true;
    return $value;
}
```
and call like this:
```php
$value = $this->unserialize($value, $isHit);
```

kynx replied:
Can do. Some people (like prophesy) hate pass-by-reference. I'm agnostic :)

marc-mabe replied:
Yes, pass-by-reference is not the cleanest approach but nowadays IDEs automatically detect that. Returning an associative array and using `extract()` on the other side is much more wired and unreadable.

marc-mabe commented:
@kynx #80 has been merged into develop now. Please can you update your PR and not allow adapters capability lockOnExpire > 0 thanks!

samsonasik commented:
@kynx it needs rebase.

kynx commented:
@samsonasik Yes, apologies, I've been pretty busy recently. I'll look at doing that this weekend.

marc-mabe commented:
@kynx I did the following changes and merged it now:

* Disallow storages with capability "lock-on-expire" as this violates PSR-6
* calculate TTL from `DateInterval` with just one system call for the current time
* skip PSR integration tests on `ExtensionNotLoadedException`
* added PSR integration test for APCu adapter

Thanks for your work !

kynx commented:
@marc-mabe Excellent! Thanks for the cleanup, and sorry I faded away. Glad to see this merged!
